### PR TITLE
wrap/unwrap ETH pattern

### LINKED ISSUE #334: lpWPowerPerpPool sometimes is used where weth is wrapped before calling othertimes requries _params.ethAmount
- in ```rebalanceWithoutVault``` any eth attached is wrapped to weth, calling with ```_params.ethAmount``` > 0 will fail
- in ```rebalanceVaultNft``` can have a mix of weth and eth (weth when reducing liquidity, eth from msg.value, removing collateral). correctly specifying the ethAmount seems tricky and its possible that we will have a mix of weth and eth that will need to be tracked and correctly passed (ethAmount is the msg.value and amount0Desired/amount1Desired would be weth and eth combined (can it use a mix?) )

### PR DESCRIPTION
**Pull request by haythem96**
# wrap/unwrap ETH pattern

## Description

This PR will make sure only WETH is used during the execution of any of the ControllerHelper functionalities, it only unwrap to ETH and use it when depositing collateral into vault or when sending back funds to user at the end of the transaction.

- Wrap any sent ETH to WETH as fist step in any payable function
- Wrap any removed ETH from vault to WETH
- Unwrap WETH before any collateral deposit to vault
- Unwrap all WETH balance at the end of transaction to send back ETH to user

Fixes #334 #357 #358 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

### CODE REVIEW

File: packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol [hunk 1]
```diff
@@ -71,7 +71,9 @@ library ControllerHelperUtil {
      * @param _isWethToken0 bool variable indicate if Weth token is token0 in Uniswap v3 weth/wPowerPerp pool
```
Comment (alpinechicken) [inline packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol:71]: add @param _weth

File: packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol [hunk 2]
```diff
@@ -138,7 +140,9 @@ library ControllerHelperUtil {
      * @param __wPowerPerpToMint amount of wPowerPerp to mint
      * @param _collateralToDeposit amount of collateral to deposit
      */
-    function mintIntoVault(address _controller, uint256 _vaultId, uint256 __wPowerPerpToMint, uint256 _collateralToDeposit) public returns (uint256) {
+    function mintIntoVault(address _controller, address _weth, uint256 _vaultId, uint256 __wPowerPerpToMint, uint256 _collateralToDeposit) public returns (uint256) {
```
Comment (alpinechicken) [inline packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol:143]: add @param _weth

File: packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol [hunk 3]
```diff
@@ -138,7 +140,9 @@ library ControllerHelperUtil {
      * @param __wPowerPerpToMint amount of wPowerPerp to mint
```
Comment (alpinechicken) [inline packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol:140]: intentional __?

**Review by alpinechicken — APPROVED** [review]
lgtm

**Review by aleone — APPROVED** [review]
lgtm, 2 issues made related to things saw in here, can add in this pr or in separate ones. nice clean refactor :)

**Comment by haythem96** [comment]
> lgtm, 2 issues made related to things saw in here, can add in this pr or in separate ones. nice clean refactor :)

Fixed the 2 new opened issues in this PR also

**Review by aleone — APPROVED** [review]

**Comment by aleone** [comment]
> > lgtm, 2 issues made related to things saw in here, can add in this pr or in separate ones. nice clean refactor :)

Fixed the 2 new opened issues in this PR also 

great!

**Pull request merged**